test/expr/closure/implicit_weak_to_strong_captures.swift

```diff
@@ -0,0 +1,94 @@
+// RUN: %target-typecheck-verify-swift -disable-availability-checking
+
+func escape(_ closure: @escaping () -> Void) {}
+func noescape(_ closure: () -> Void) {}
+
+class SomeClass {}
+
+class TestImplicitWeakToStrongCaptures {
+  func test_self() {
+    escape { // expected-note {{'self' implicitly captured here}}
+      escape { [weak self] in  // expected-warning {{'weak' capture list item 'self' implicitly captured as strong reference in outer scope}}
+        _ = self
+      }
+    }
+  }
+
+  func test_locals() {
+    let a = SomeClass()
+    escape { // expected-note {{'a' implicitly captured here}}
+      let b = SomeClass()
+      escape { [
+        weak a, // expected-warning {{'weak' capture list item 'a' implicitly captured as strong reference in outer scope}}
+        weak b
+      ] in
+        _ = a
+        _ = b
+      }
+    }
+  }
+
+  func test_flavors_of_weak() {
+    escape {  // expected-note 3 {{'self' implicitly captured here}}
+      escape { [weak self] in _ = self }    // expected-warning {{'weak' capture list item 'self' implicitly captured as strong reference in outer scope}}
+      escape { [unowned self] in _ = self } // expected-warning {{'unowned' capture list item 'self' implicitly captured as strong reference in outer scope}}
+      escape { [unowned(unsafe) self] in _ = self } // expected-warning {{'unowned(unsafe)' capture list item 'self' implicitly captured as strong reference in outer scope}}
+    }
+  }
+
+  func test_binding_new_name() {
+    escape {  // expected-note {{'self' implicitly captured here}}
+      escape { [weak welf = self] in _ = welf }    // expected-warning {{'weak' capture list item 'self' implicitly captured as strong reference in outer scope}}
```

Comment:
this raises a question about wording – in the case where the binding introduces a new variable name, the logic currently refers to the referent's name (and source loc) in the diagnostic output; the new variable's name isn't used at all. is this confusing?

Reply:
Maybe it could be re-worded as `'weak' capture implicitly captures 'self' as strong reference in outer scope`? That way it's clear it's referring to the source of the capture, and you could augment it to be `'weak' capture 'welf' implicitly captures 'self' as strong reference in outer scope` in cases where the bound name differs?